breakpad-symbols/src/lib.rs

```diff
@@ -310,12 +310,23 @@ impl HttpSymbolSupplier {
 
 /// Save the data in `contents` to `path`.
 fn save_contents(contents: &[u8], path: &Path) -> io::Result<()> {
+    // Use tempfile to save things to our cache to ensure proper
+    // atomicity of writes. We may want multiple instances of rust-minidump
+    // to be sharing a cache, and we don't want one instance to see another
+    // instance's partially written results.
+    //
+    // tempfile is designed explicitly for this purpose, and will handle all
+    // the platform-specific details and do its best to cleanup if things
+    // crash. See `std::env::temp_dir()` for how it chooses where to put
+    // the temporary file (defaults to /tmp on linux).
     let base = path.parent().ok_or_else(|| {
         io::Error::new(io::ErrorKind::Other, format!("Bad cache path: {:?}", path))
     })?;
     fs::create_dir_all(&base)?;
-    let mut f = File::create(path)?;
-    f.write_all(contents)?;
+
+    let mut temp = tempfile::NamedTempFile::new()?;
+    temp.write_all(contents)?;
+    temp.persist(path)?;
```

Comment:
I think this doesn't work with the way we've got Socorro processor set up. Let me do a brain dump....

We run the Socorro processor in a Docker container with a volume mount for the disk cache which has both the `symbols-cache` and `symbols-tmp` directories in it. Thus those directories aren't on the same mount as `/tmp`.

The `symbols-cache` directory is managed by the disk cache manager which is a separate process that runs in the container and uses inotify to watch for changes in the directory and maintains the files in a LRU fashion. We don't want it watching `symbols-tmp` because then partially-written files kick off inotify events and cache updates and evictions and everything gets harder to reason about. Fun times.

So what we want is something like:

1. The downloader creates a temp file with part of the filename random to write the bytes of the sym file into. The random part asserts some minor security goodness. I think it also makes the code writing the file easier to deal with since we can ignore the "what if something else is already writing this temp file but the final file isn't there?" scenario.
2. Once the temp file is completely written out and closed, the downloader tries to rename the temp file to the destination final file name. Because it's on the same volume, the downloader can do a rename which (I'm fairly sure) is atomic. This "moves" the file from `symbols-tmp` to `symbols-cache`. The disk cache manager will see the file (inotify ftw) and update the LRU and evict things as appropriate. If there's already a file at the destination, the downloader can be a little sad it did all that work, delete the temp file, and move on with life--everything is in a known good state.

Given that, I think we want to pass along the `symbols_tmp_path` and use `new_in`. Also, I think we want to use `persist_noclobber` and then I think we should be good. We need to pass along the `symbols-tmp` path and then tweak these three lines like this (barring mediocre Rust code):

```rust

let mut temp = tempfile::NamedTempFile::new_in(symbols_tmp_path)?;
temp.write_all(contents)?;
temp.persist_noclobber(path)?;
```

If any of this fails, then that's fine--we can still use the sym file we downloaded. It just doesn't get cached. Since each subsequent line depends on the success of the previous lines, we might want to check and just drop out if we hit a failure.

Having said that, the processor runs this as a subprocess, so it this fails and the stackwalk process dies, the processor will make note of that in the processed crash report and move on.

Having said that, while I keep track of exit codes in Grafana, I'm not currently looking at the `processor_notes` field because it's a mess and hard to track in aggregate. That's covered in [bug 1626811](https://bugzilla.mozilla.org/show_bug.cgi?id=1626811).

I think that's all correct. If it's not, it'd be great to be corrected because I'm basing a few things on this mental model of the world.